File: src/main/java/com/mooregreatsoftware/gradle/defaults/config/LicenseConfig.java
/*
 * Copyright 2014-2016 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooregreatsoftware.gradle.defaults.config;

import groovy.lang.GroovyObject;
import lombok.val;
import nl.javadude.gradle.plugins.license.License;
import nl.javadude.gradle.plugins.license.LicenseExtension;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.gradle.api.Project;
import org.gradle.api.plugins.Convention;

import java.util.function.Supplier;

@SuppressWarnings("WeakerAccess")
public class LicenseConfig extends AbstractConfig {
    public LicenseConfig(Project project) {
        super(project);
    }


    public void config(Supplier<@Nullable String> copyrightYears) {
        plugins().apply("license");
        final LicenseExtension licenseExt = project.getConvention().getByType(LicenseExtension.class);
        licenseExt.setHeader(project.getRootProject().file("gradle/HEADER"));
        licenseExt.setStrictCheck(true);
        licenseExt.setUseDefaultMappings(false);
        licenseExt.mapping("groovy", "SLASHSTAR_STYLE");
        licenseExt.mapping("java", "SLASHSTAR_STYLE");

        project.afterEvaluate(prj -> configLicenseExtension(copyrightYears, licenseExt));

        tasks().whenTaskAdded(task -> {
            if (task instanceof License) {
                ((License)task).exclude("**/*.properties");
            }
        });
    }


    private static void configLicenseExtension(Supplier<@Nullable String> copyrightYears, LicenseExtension licenseExt) {
        @SuppressWarnings("RedundantCast") val metaClass = ((GroovyObject)licenseExt).getMetaClass();
        val conExtPropsProp = metaClass.getMetaProperty("extensions");
        val convExtensions = (Convention)conExtPropsProp.getProperty(licenseExt);
        val extensions = convExtensions.getExtraProperties();
        val years = copyrightYears.get();
        if (years != null) extensions.set("year", years);
    }

}

File: src/main/kotlin/com/mooregreatsoftware/gradle/defaults/config/LicenseConfig.kt
/*
 * Copyright 2014-2016 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooregreatsoftware.gradle.defaults.config

import groovy.lang.GroovyObject
import nl.javadude.gradle.plugins.license.LicenseExtension
import nl.javadude.gradle.plugins.license.LicensePlugin
import org.gradle.api.Project
import org.gradle.api.plugins.Convention
import org.gradle.api.plugins.ExtraPropertiesExtension
import java.util.function.Supplier

class LicenseConfig(private val project: Project) {

    fun config(copyrightYears: Supplier<String>) {
        project.plugins.apply(LicensePlugin::class.java)
        val licenseExt = project.convention.getByType(LicenseExtension::class.java)
        licenseExt.header = project.rootProject.file("gradle/HEADER")
        licenseExt.strictCheck = true
        licenseExt.isUseDefaultMappings = true
        licenseExt.mapping("groovy", "SLASHSTAR_STYLE")
        licenseExt.mapping("java", "SLASHSTAR_STYLE")
        licenseExt.mapping("scala", "SLASHSTAR_STYLE")
        licenseExt.mapping("kt", "SLASHSTAR_STYLE")
        licenseExt.mapping("css", "SLASHSTAR_STYLE")

        project.afterEvaluate { prj -> configLicenseExtension(copyrightYears, licenseExt) }
    }


    private fun configLicenseExtension(copyrightYears: Supplier<String>, licenseExt: LicenseExtension) {
        licenseExt.exclude("**/*.properties")

        val years = copyrightYears.get()
        if (years != null) licenseExt.ext.set("year", years)
    }

}

val LicenseExtension.ext: ExtraPropertiesExtension
    get() {
        val metaClass = (this as GroovyObject).metaClass
        val conExtPropsProp = metaClass.getMetaProperty("extensions")
        val convExtensions = conExtPropsProp.getProperty(this) as Convention
        return convExtensions.extraProperties
    }
